### PR TITLE
Updated frontend: implemented "Save" for CPU and Heap profiles.

### PR DESCRIPTION
**Pull request by marcominetti**

### CODE REVIEW

File: front-end-node/profiler/SaveOverrides.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  var url = URL.createObjectURL(blob);
+  var link = document.createElementNS("http://www.w3.org/1999/xhtml", "a");
+  link.download = filename;
+  link.href = url;
+  var event = document.createEvent("MouseEvents");
+  event.initMouseEvent(
+    "click", true, false, window, 0, 0, 0, 0, 0
+    , false, false, false, false, 0, null
+  );
+  link.dispatchEvent(event);
```
Comment (3y3) [inline front-end-node/profiler/SaveOverrides.js]: How about simple `link.click()` here?
Or it will not work in some cases?

Comment (marcominetti) [inline front-end-node/profiler/SaveOverrides.js]: I saw it in FileSaver but it's deprecated according to MDN https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/initMouseEvent

I'll try `link.click()` or switch to proper event dispatching for Chromium project...

File: front-end-node/profiler/SaveOverrides.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    self._fileName = self._fileName || new Date().toISO8601Compact() + self._profileType.fileExtension();
+    if (self._tempFile == null) {
+        self._oldOnTempFileReady = self._onTempFileReady;
+        self._onTempFileReady = function () {
+            self._tempFile.read(function (data) {
+                WebInspector.saveAs(new Blob([data], { type: "application/octet-stream" }), self._fileName);
+            });
+            self._onTempFileReady = self._oldOnTempFileReady;
+            if (self._onTempFileReady) {
+                self._onTempFileReady();
```
Comment (3y3) [inline front-end-node/profiler/SaveOverrides.js]: Looking on algorithm implemented by devtools I don't see here a situation, when we need to call `_oldOnTempFileReady`. Correct me if I'm wrong.

File: front-end-node/profiler/SaveOverrides.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+                self._onTempFileReady = null;
+            }
+        };
+    } else {
+        self._tempFile.read(function (data) {
+            WebInspector.saveAs(new Blob([data], { type: "application/octet-stream" }), self._fileName);
+        });
```
Comment (3y3) [inline front-end-node/profiler/SaveOverrides.js]: You use equal code in some places. It well be better to move it to new function.

File: front-end-node/profiler/SaveOverrides.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  var url = URL.createObjectURL(blob);
+  var link = document.createElementNS("http://www.w3.org/1999/xhtml", "a");
+  link.download = filename;
+  link.href = url;
+  link.click();
+  URL.revokeObjectURL(url);
+};
+
+WebInspector.CPUProfileHeader.prototype.saveToFile = WebInspector.HeapProfileHeader.prototype.saveToFile = function() {
+    var self = this;
```
Comment (3y3) [inline front-end-node/profiler/SaveOverrides.js]: We can avoid closure of context

**Comment by 3y3** [comment]
Please look on my inline comments. I propose next implementation of `saveToFile` that fixes all my comments, but I don't test this code:
```js
WebInspector.CPUProfileHeader.prototype.saveToFile = 
WebInspector.HeapProfileHeader.prototype.saveToFile = function() {
  var fileName = this._fileName || new Date().toISO8601Compact() + this._profileType.fileExtension();

  this._onTempFileReady = function() {
    this._tempFile.read(function (data) {
      WebInspector.saveAs(new Blob([data], {type: 'application/octet-stream'}), fileName);
    });

    this._onTempFileReady = null;
  };
  
  if (this._tempFile) this._onTempFileReady();
};
```

That you think about this realisation? Is it working as expected?

**Comment by marcominetti** [comment]
I think you're going to save on every profile completed... I test it just now...
In case another option could be 

```javascript
WebInspector.CPUProfileHeader.prototype.saveToFile =
WebInspector.HeapProfileHeader.prototype.saveToFile = function () {
    var fileName = this._fileName || new Date().toISO8601Compact() + this._profileType.fileExtension();

    function readAndSave(ProfileHeader){
      ProfileHeader._tempFile.read(function (data) {
          WebInspector.saveAs(new Blob([data], {type: "application/octet-stream"}), fileName);
      });
    }

    if (this._tempFile == null) {
        this._oldOnTempFileReady = this._onTempFileReady;
        this._onTempFileReady = (function () {
            readAndSave(this);
            this._onTempFileReady = self._oldOnTempFileReady;
            if (this._onTempFileReady) {
                this._onTempFileReady();
                this._onTempFileReady = null;
            }
        }).bind(this);
    } else {
        readAndSave(this);
    }
}
```

**Comment by marcominetti** [comment]
Updated and squashed.

File: front-end-node/profiler/SaveOverrides.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
+/*jshint browser:true, nonew:false*/
+/*global WebInspector:true*/
+
+WebInspector.saveAs = function(blob,filename) {
```
Comment (3y3) [inline front-end-node/profiler/SaveOverrides.js]: Please add space after comma.

File: front-end-node/profiler/SaveOverrides.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  var url = URL.createObjectURL(blob);
+  var link = document.createElementNS("http://www.w3.org/1999/xhtml", "a");
```
Comment (3y3) [inline front-end-node/profiler/SaveOverrides.js]: Linter fails here. Run `npm test` to see the result or check in here [travisci](https://travis-ci.org/node-inspector/node-inspector/builds/59524079)
Don't worry about all tests here, but linter errors is in scope of this pr.

**Comment by 3y3** [comment]
I added some inline comments. They are last before merging. You can publish here already squashed version for this pr.

Thank you for all front-end fixes. I'll start to review your features.

**Comment by marcominetti** [comment]
Done ;)

**Comment by 3y3** [comment]
Landed as 99514aa3d211242505c6cf95da1e2ef953023c8f

**Pull request closed without merging**